src/main/java/net/minecraftforge/client/ForgeHooksClient.java

```diff
@@ -565,6 +576,152 @@ public static void registerTESRItemStack(Item item, int metadata, Class<? extend
+                skyLight += (int) ((quadData[lmapIndex][i][1] * 0xFFFF) / 0x20);
+            }
+            // Values must be multiplied by 16, divided by 4 for average => x4
+            blockLight *= 4;
+            skyLight *= 4;
+        }
+        
+        // Dummy overrides
+
+        @Override
+        public void setQuadTint(int tint) {}
+
+        @Override
+        public void setQuadOrientation(EnumFacing orientation) {}
+
+        @Override
+        public void setApplyDiffuseLighting(boolean diffuse) {}
+
+        @Override
+        public void setTexture(TextureAtlasSprite texture) {}
+    }
+    
+    private static final LightGatheringTransformer lightGatherer = new LightGatheringTransformer();
+
+    public static void renderLitItem(RenderItem ri, IBakedModel model, int color, ItemStack stack) 
+    {
+        BufferBuilder bufferbuilder = Tessellator.getInstance().getBuffer();
+
+        List<BakedQuad> allquads = new ArrayList<>();
+
+        for (EnumFacing enumfacing : EnumFacing.VALUES)
+        {
+            allquads.addAll(model.getQuads(null, enumfacing, 0));
+        }
+
+        allquads.addAll(model.getQuads(null, null, 0));
+
+        // Current list of consecutive quads with the same lighting
+        List<BakedQuad> segment = new ArrayList<>();
+
+        // Lighting of the current segment
+        int segmentBlockLight = -1;
+        int segmentSkyLight = -1;
+
+        for (int i = 0; i < allquads.size(); i++) 
+        {
+            BakedQuad q = allquads.get(i);
+
+            // Lighting of the current quad
+            int bl = 0;
+            int sl = 0;
+
+            // Fail-fast on ITEM, as it cannot have light data
+            if (q.getFormat() != DefaultVertexFormats.ITEM) 
+            {
+                lightGatherer.setVertexFormat(q.getFormat());
+                if (lightGatherer.hasLighting())
+                {
+                    q.pipe(lightGatherer);
+                }
+                bl = lightGatherer.blockLight;
+                sl = lightGatherer.skyLight;
+            }
+
+            // If this is a new light value, draw the segment and flush it
+            if (segmentBlockLight != bl || segmentSkyLight != sl) 
+            {
+                if (i > 0) // Make sure this isn't the first quad being processed 
+                {
+                    drawSegment(bufferbuilder, ri, color, stack, segment, segmentBlockLight, segmentSkyLight, true);
+                }
+                segmentBlockLight = bl;
+                segmentSkyLight = sl;
+            }
+
+            segment.add(q);
+        }
+        
+        boolean lightDirty = segment.size() < allquads.size(); // Was at least one change in lighting for this model
+        drawSegment(bufferbuilder, ri, color, stack, segment, segmentBlockLight, segmentSkyLight, lightDirty);
+
+        // Clean up render state if necessary
+        if (lightDirty)
```

Comment:
Does this logic need changing now - the render state used by the final segment might be without emissive lighting, in which case this is setting the same state twice.